addons/account/partner.py

```diff
@@ -241,11 +241,32 @@ def _debit_search(self, cr, uid, obj, name, args, context=None):
     def _invoice_total(self, cr, uid, ids, field_name, arg, context=None):
         result = {}
         account_invoice_report = self.pool.get('account.invoice.report')
-        for partner in self.browse(cr, uid, ids, context=context):
-            domain = [('partner_id', 'child_of', partner.id), ('state', 'not in', ['draft', 'cancel'])]
+        user = self.pool['res.users'].browse(cr, uid, uid, context=context)
+        for partner_id in ids:
+            # first read invoices in the user's currency
+            all_partner_ids = self.pool['res.partner'].search(
+                cr, uid, [('id', 'child_of', partner_id)], context=context)
+            # searching account.invoice.report via the orm is comparatively
+            # expensive (searching for many2one fields generates in $idlist
+            # queries), so we do the simplest and most common case in sql
+            cr.execute(
+                'select sum(price_total) '
+                'from account_invoice_report '
+                'where '
+                'partner_id in %s and state not in %s and currency_id=%s',
+                (tuple(all_partner_ids), tuple(['draft', 'cancel']),
+                 user.company_id.currency_id.id))
+            result[partner_id] = sum(s or 0.0 for s, in cr.fetchall())
+            # and then add invoices in other currencies
+            domain = [
+                ('partner_id', 'in', all_partner_ids),
+                ('state', 'not in', ['draft', 'cancel']),
+                ('currency_id', '!=', user.company_id.currency_id.id),
+            ]
             invoice_ids = account_invoice_report.search(cr, uid, domain, context=context)
-            invoices = account_invoice_report.browse(cr, uid, invoice_ids, context=context)
-            result[partner.id] = sum(inv.user_currency_price_total for inv in invoices)
+            if invoice_ids:
```

Comment:
this is for invoices not in the user's currency. For them, it's a lot more convenient to read it through the view and get the currency conversions for free.